src/nav/InitialNavigationDispatcher.js

```diff
@@ -0,0 +1,83 @@
+/* @flow strict-local */
+import type { Node as React$Node } from 'react';
+import { PureComponent } from 'react';
+
+import type { Dispatch, Account, User } from '../types';
+import { resetToAccountPicker, resetToRealmScreen, resetToMainTabs } from '../actions';
+import { connect } from '../react-redux';
+import { getIsHydrated, hasAuth as getHasAuth } from '../selectors';
+
+type SelectorProps = $ReadOnly<{|
+  isHydrated: boolean,
+  hasAuth: boolean,
+  accounts: Account[],
+  users: User[],
+|}>;
+
+type Props = $ReadOnly<{|
+  children: React$Node,
+
+  dispatch: Dispatch,
+  ...SelectorProps,
+|}>;
+
+class InitialNavigationDispatcher extends PureComponent<Props> {
+  componentDidMount() {
+    if (this.props.isHydrated) {
+      this.doInitialNavigation();
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    if (!prevProps.isHydrated && this.props.isHydrated) {
+      this.doInitialNavigation();
+    }
+  }
+
+  /**
+   * Data has been loaded, so open the app to the right screen.
+   *
+   * Not to be called before the REHYDRATE action, and not to be
+   * called more than once.
+   */
+  doInitialNavigation = () => {
+    const { hasAuth, accounts, users, dispatch } = this.props;
+
+    // If there are accounts but the active account is not logged in,
+    // show account screen.
+    if (!hasAuth) {
```

Comment:
The "If there are accounts" part no longer applies -- in the old code, it's there to make a contrast with the preceding condition:
```js
  // If there's no data to rehydrate, or no account data, show login screen.	
  if (!action.payload || !action.payload.accounts) {	
    return getStateForRoute('realm');	
  }	
```
but that's dissolved by the fact that with this design we're looking at a nice well-formed state post-REHYDRATE, rather than the REHYDRATE payload which is a bunch of pieces in roughly the shape of our state.

Reply:
You're saying the comment should be adjusted, right:

```diff
-   // If there are accounts but the active account is not logged in,
-   // show account screen.
+   // If the active account is not logged in, show account screen.
```

Reply:
I've stared at this a bit, and I think something is still off:

```js
    // If the active account is not logged in, show account screen.
    if (!hasAuth) {
      if (accounts.length > 1) {
        dispatch(resetToAccountPicker());
        return;
      } else {
        dispatch(resetToRealmScreen({ initial: true }));
        return;
      }
    }
```

I think maybe

```diff
-     if (accounts.length > 1) {
+     if (accounts.length >= 1) {
```

?

Reply:
Still doesn't quite line up with the comment; maybe something like this, to replace the code in the multi-line quote [above](https://github.com/zulip/zulip-mobile/pull/4274#discussion_r508171639):

```js
    // If we don't know about any accounts, go to the realm-input screen.
    if (accounts.length === 0) {
      dispatch(resetToRealmScreen({ initial: true }));
      return;
    }

    // If the active account is not logged in, show account screen.
    if (!hasAuth) {
      dispatch(resetToAccountPicker());
      return;
    }
```

Reply:
(Discussed [in chat](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/decouple.20nav.20from.20redux.20%28.23M3804%29/near/1042584), and the result is in your new tip commit.)